Code/GraphMol/Sgroup.h

```diff
@@ -0,0 +1,215 @@
+//
+//
+//  Copyright (C) 2002-2018 Greg Landrum and T5 Informatics GmbH
+//
+//   @@ All Rights Reserved @@
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+/*! \file Sgroup.h
+
+  \brief Defines the SGroup class
+
+*/
+#include <RDGeneral/export.h>
+#ifndef _RD_SGROUP_H
+#define _RD_SGROUP_H
+
+#include <unordered_map>
+
+#include <Geometry/point.h>
+#include <RDGeneral/types.h>
+#include <boost/smart_ptr.hpp>
+
+namespace RDKit {
+class ROMol;
+class Bond;
+class Atom;
+
+//! used to indicate errors from incorrect sgroup access
+class RDKIT_GRAPHMOL_EXPORT SGroupException : public std::runtime_error {
+ public:
+  //! construct with an error message
+  SGroupException(const char *msg) : std::runtime_error(msg){};
+  //! construct with an error message
+  SGroupException(const std::string &msg) : std::runtime_error(msg){};
+};
+
+class RDKIT_GRAPHMOL_EXPORT SGroup {
+ public:
+  enum class BondType {
+    XBOND,  // External
```

Comment:
I prefer explicitness, perhaps we could use

```
EXTERNAL, // XBOND
INTERNAL, //CBOND
```

But this isn't crucial.

Reply:
It might also help error messages
```
 errout << "Bond with index " << bond->getIdx()
            << " is not an external bond for SGroup " << getId();
```

Reply:
I agree on the error message, but I actually prefer the XBOND/CBOND names, as these are the labels used in the SD specification. I'm not adamant on that, though.